text/2022-04-29-tls-support-for-httpchaos.md

```diff
@@ -0,0 +1,110 @@
+# TLS Support for HTTPChaos
+
+## Summary
+
+A solution to support TLS in HTTPChaos.
+
+## Motivation
+
+Currently, our HTTPChaos can only injects faults into unencrypted TCP connection.
+Once the clients and servers connect to others by TLS, we can do nothing in the middle.
+With the HTTPS protects more and more connections, we must figure out a solution to attack it.
+
+## Detailed Design
+
+### User-Provided Secrets
+
+The simplest solutions, mainly used in the case of server-side HTTPChaos, is providing private-key(the secret) by the user.
+In server-side HTTPChaos, we should inject some inbound stream of a server, so the user should be willing
+to provide private-key of the certificate. In this solution, our chaos-tproxy can decrypt the client-generated
+symmetric-key by the private-key.
+
+Here is the schematic diagram:
+
+```
+
+STEP
+
+               request for tls connection                     request for tls connection
+1.  client  <------------------------------>  chaos-tproxy <------------------------------> server
+                    cert(public-key)          (private-key)         cert(public-key)        (private-key)
+
+             encrypted(by public-key) symmetric-key                 encrypted(by public-key) symmetric-key
+2.  client -----------------------------------------> chaos-tproxy ---------------------------------------> server
+                                                (decrypted symmetric-key)                           (decrypted symmetric-key)
+
+             encrypted(by symmetric-key) request data                           modifited request data
+3.  client <------------------------------------------> chaos-tproxy <--------------------------------------------> server
+                    modifited response data           (decrypted data)  encrypted(by symmetric-key) response data 
+```
+
+The user can also provide different certificates from the server, so long as the certificate is trusted by the client.
+
+For example, we can configure the secrets in HTTPChaos:
+
+```yaml
+apiVersion: chaos-mesh.org/v1alpha1
+kind: HTTPChaos
+spec:
+  ...
+  tls:
+    secrets:
+      privateKey: <the secretName of private-key>
+      certificate: <the secretName of certificate>
+
+```
```

Comment:
Well, if we assume that, the user could always provide the certs which are trusted by both "client" and "server"(with mtls), I think it could be easily applied on both server-side and client-side injection.

> the user could always provide the certs which are trusted by both "client" and "server"

And actually, I think it's not a kind of difficult for any companies/orgs which hold their own PKI infrastructure, they require tons of certs with TLS for encrypting their internal/public networking communication/ service RPC between applications.

Reply:
Yes